pkg/cli/cli.go

```diff
@@ -0,0 +1,498 @@
+		}
+	} else if err == nil {
+		c.configured = true
+		c.projectVersion = projectConfig.Version
+	} else {
+		return fmt.Errorf("failed to read config: %v", err)
+	}
+
+	// Validate after setting projectVersion but before buildRootCmd so we error
+	// out before an error resulting from an incorrect cli is returned downstream.
+	if err = c.validate(); err != nil {
+		return err
+	}
+
+	// When invoking 'init', a user can:
+	// 1. Not set --plugins
+	// 2. Set --plugins to a plugin, ex. --plugins=go-x
+	// In case 1, default plugins will be used to determine which plugin to use.
+	// In case 2, the value passed to --plugins is used.
+	// For all other commands, a config's 'layout' key is used. Since both
+	// layout and --plugins values can be short (ex. "go/v2.0.0") or unversioned
+	// (ex. "go.kubebuilder.io") keys or both, their values may need to be
+	// resolved to known plugins by key.
+	plugins := c.pluginsFromOptions[c.projectVersion]
+	switch {
+	case c.cliPluginKey != "":
+		// Filter plugin by keys passed in CLI.
+		c.resolvedPlugins, err = resolvePluginsByKey(plugins, c.cliPluginKey)
+	case c.configured && projectConfig.IsV3():
+		// All non-v1 configs must have a layout key. This check will help with
+		// migration.
+		if projectConfig.Layout == "" {
+			return fmt.Errorf("config must have a layout value")
+		}
+		// Filter plugin by config's layout value.
+		c.resolvedPlugins, err = resolvePluginsByKey(plugins, projectConfig.Layout)
+	default:
+		// Use the default plugins for this project version.
+		c.resolvedPlugins = c.defaultPluginsFromOptions[c.projectVersion]
+	}
+	if err != nil {
+		return err
+	}
+
+	c.cmd = c.buildRootCmd()
+
+	// Add extra commands injected by options.
+	for _, cmd := range c.extraCommands {
+		for _, subCmd := range c.cmd.Commands() {
+			if cmd.Name() == subCmd.Name() {
+				return fmt.Errorf("command %q already exists", cmd.Name())
+			}
+		}
+		c.cmd.AddCommand(cmd)
+	}
+
+	// Write deprecation notices after all commands have been constructed.
+	for _, p := range c.resolvedPlugins {
+		if d, isDeprecated := p.(plugin.Deprecated); isDeprecated {
+			fmt.Printf(noticeColor, fmt.Sprintf("[Deprecation Notice] %s\n\n",
+				d.DeprecationWarning()))
+		}
+	}
+
+	return nil
+}
+
+// parseBaseFlags parses the command line arguments, looking for flags that
+// affect initialization of a cli. An error is returned only if an error
+// unrelated to flag parsing occurs.
+func (c *cli) parseBaseFlags() error {
+	// Create a dummy "base" flagset to populate from CLI args.
+	fs := pflag.NewFlagSet("base", pflag.ExitOnError)
+	fs.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
+
+	var help bool
+	// Set base flags that require pre-parsing to initialize c.
+	fs.BoolVarP(&help, helpFlag, "h", false, "print help")
+	fs.StringVar(&c.projectVersion, projectVersionFlag, c.defaultProjectVersion, "project version")
+	fs.StringVar(&c.cliPluginKey, pluginsFlag, "", "plugins to run")
```

Comment:
Is this plural for forward-compatibility reasons? Should this use `fs.StringSliceVar()`?

For now, should we check to ensure only 1 plugin is provided?

Reply:
For forward compatibility reasons yes. I have it as a string so users aren't confused about what type (string vs string list) they can pass to `--plugins`. Once phase 2 is done we can update the flag type without actually breaking the flag API.